### DISCUSSION TITLE
[FR] support Lit SSR

### DISCUSSION THREAD

**Discussion by e111077** [ing-bank/lion#1611] [Ideas]
Heya, I've been working on SSR-ing lit.dev and we use lion combobox, but we've run into a few issues while attempting to SSR lion-wc. In particular:

* WC definitions must be loaded in Node js so they must adhere to Node's requirements for ES modules
  * package.json must define `"type": "module"` or JS files must use `.mjs` extension
  * adding `"type": "module"` only to ssr-compatible components might be a good way to advertise "this is SSR ready"
* Complex DOM calls should not exist in `connectedCallback` `willUpdate` and `constructor`
  * I found this to happen in browser capability checks and polyfills that run at import or constructor
  * 100% happy to look into any SSR WebComponent polyfill issues y'all run into
  * There might be some issues regarding scoped elements here.

I've spun up a [super simple lit SSR project on glitch](https://glitch.com/edit/#!/lit-ssr-experimental?path=ssr.js%3A1%3A0), but I highly recommend downloading the project locally as glitch has some weird problems specific to node because of its online environment. Also, tried stackblitz, but it runs such a unique custom implementation of Node 13 (we are on 17 now) so that was generally untenable.

Happy to share any domain knowledge on the topic!

**Comment by daKmoR** [comment]
We definitely want to support SSR... sadly it's not high on the priority list now 🙈

I expect it to be quite some work
- [ssr] we do a lot of "rendering" to light dom for accessibility (which is currently not supported in SSR as far as I can tell)
- [lion] we do the light dom rendering during connectedCallback (might need some different approach here)
- [ssr] we probably need a dedicated "ScopedElements" renderer to discover custom elements that need to be considered during SSR
- [lion] we need to get rid of some of the inlined polyfills (already on our TODO as we dropped IE)
- [lion] we need to get rid of browser only side effects (yes please)
- [lion] we need to set the `"type": "module"`
- [lion] I'm unsure if any work will be needed for the overlay system to support SSR 🤔

To summarize I think there are like two things we would need to improve on SSR and a lot more on the lion code itself...
